// fast-uri/lib/schemes.js
'use strict'

const { isUUID } = require('./utils')
const URN_REG = /([\da-z][\d\-a-z]{0,31}):((?:[\w!$'()*+,\-.:;=@]|%[\da-f]{2})+)/iu

const supportedSchemeNames = /** @type {const} */ (['http', 'https', 'ws',
  'wss', 'urn', 'urn:uuid', 'mailto'])

/** @typedef {supportedSchemeNames[number]} SchemeName */

/**
 * @param {string} name
 * @returns {name is SchemeName}
 */
function isValidSchemeName (name) {
  return supportedSchemeNames.indexOf(/** @type {*} */ (name)) !== -1
}

/**
 * @callback SchemeFn
 * @param {import('../types/index').URIComponent} component
 * @param {import('../types/index').Options} options
 * @returns {import('../types/index').URIComponent}
 */

/**
 * @typedef {Object} SchemeHandler
 * @property {SchemeName} scheme - The scheme name.
 * @property {boolean} [domainHost] - Indicates if the scheme supports domain hosts.
 * @property {SchemeFn} parse - Function to parse the URI component for this scheme.
 * @property {SchemeFn} serialize - Function to serialize the URI component for this scheme.
 * @property {boolean} [skipNormalize] - Indicates if normalization should be skipped for this scheme.
 * @property {boolean} [absolutePath] - Indicates if the scheme uses absolute paths.
 * @property {boolean} [unicodeSupport] - Indicates if the scheme supports Unicode.
 * @property {boolean} [skipEscape] - Indicates if generic path escaping should be skipped for this scheme.
 */

/**
 * @param {import('../types/index').URIComponent} wsComponent
 * @returns {boolean}
 */
function wsIsSecure (wsComponent) {
  if (wsComponent.secure === true) {
    return true
  } else if (wsComponent.secure === false) {
    return false
  } else if (wsComponent.scheme) {
    return (
      wsComponent.scheme.length === 3 &&
      (wsComponent.scheme[0] === 'w' || wsComponent.scheme[0] === 'W') &&
      (wsComponent.scheme[1] === 's' || wsComponent.scheme[1] === 'S') &&
      (wsComponent.scheme[2] === 's' || wsComponent.scheme[2] === 'S')
    )
  } else {
    return false
  }
}

/** @type {SchemeFn} */
function httpParse (component) {
  if (!component.host) {
    component.error = component.error || 'HTTP URIs must have a host.'
  }

  return component
}

/** @type {SchemeFn} */
function httpSerialize (component) {
  const secure = String(component.scheme).toLowerCase() === 'https'

  // normalize the default port
  if (component.port === (secure ? 443 : 80) || component.port === '') {
    component.port = undefined
  }

  // normalize the empty path
  if (!component.path) {
    component.path = '/'
  }

  // NOTE: We do not parse query strings for HTTP URIs
  // as WWW Form Url Encoded query strings are part of the HTML4+ spec,
  // and not the HTTP spec.

  return component
}

/** @type {SchemeFn} */
function wsParse (wsComponent) {
// indicate if the secure flag is set
  wsComponent.secure = wsIsSecure(wsComponent)

  // construct resouce name
  wsComponent.resourceName = (wsComponent.path || '/') + (wsComponent.query ? '?' + wsComponent.query : '')
  wsComponent.path = undefined
  wsComponent.query = undefined

  return wsComponent
}

/** @type {SchemeFn} */
function wsSerialize (wsComponent) {
// normalize the default port
  if (wsComponent.port === (wsIsSecure(wsComponent) ? 443 : 80) || wsComponent.port === '') {
    wsComponent.port = undefined
  }

  // ensure scheme matches secure flag
  if (typeof wsComponent.secure === 'boolean') {
    wsComponent.scheme = (wsComponent.secure ? 'wss' : 'ws')
    wsComponent.secure = undefined
  }

  // reconstruct path from resource name
  if (wsComponent.resourceName) {
    const [path, query] = wsComponent.resourceName.split('?')
    wsComponent.path = (path && path !== '/' ? path : undefined)
    wsComponent.query = query
    wsComponent.resourceName = undefined
  }

  // forbid fragment component
  wsComponent.fragment = undefined

  return wsComponent
}

/** @type {SchemeFn} */
function urnParse (urnComponent, options) {
  if (!urnComponent.path) {
    urnComponent.error = 'URN can not be parsed'
    return urnComponent
  }
  const matches = urnComponent.path.match(URN_REG)
  if (matches) {
    const scheme = options.scheme || urnComponent.scheme || 'urn'
    urnComponent.nid = matches[1].toLowerCase()
    urnComponent.nss = matches[2]
    const urnScheme = `${scheme}:${options.nid || urnComponent.nid}`
    const schemeHandler = getSchemeHandler(urnScheme)
    urnComponent.path = undefined

    if (schemeHandler) {
      urnComponent = schemeHandler.parse(urnComponent, options)
    }
  } else {
    urnComponent.error = urnComponent.error || 'URN can not be parsed.'
  }

  return urnComponent
}

/** @type {SchemeFn} */
function urnSerialize (urnComponent, options) {
  if (urnComponent.nid === undefined) {
    throw new Error('URN without nid cannot be serialized')
  }
  const scheme = options.scheme || urnComponent.scheme || 'urn'
  const nid = urnComponent.nid.toLowerCase()
  const urnScheme = `${scheme}:${options.nid || nid}`
  const schemeHandler = getSchemeHandler(urnScheme)

  if (schemeHandler) {
    urnComponent = schemeHandler.serialize(urnComponent, options)
  }

  const uriComponent = urnComponent
  const nss = urnComponent.nss
  uriComponent.path = `${nid || options.nid}:${nss}`

  options.skipEscape = true
  return uriComponent
}

/** @type {SchemeFn} */
function urnuuidParse (urnComponent, options) {
  const uuidComponent = urnComponent
  uuidComponent.uuid = uuidComponent.nss
  uuidComponent.nss = undefined

  if (!options.tolerant && (!uuidComponent.uuid || !isUUID(uuidComponent.uuid))) {
    uuidComponent.error = uuidComponent.error || 'UUID is not valid.'
  }

  return uuidComponent
}

/** @type {SchemeFn} */
function urnuuidSerialize (uuidComponent) {
  const urnComponent = uuidComponent
  // normalize UUID
  urnComponent.nss = (uuidComponent.uuid || '').toLowerCase()
  return urnComponent
}

/**
 * @param {string} value
 * @returns {string}
 */
function decodeMailtoComponent (value) {
  try {
    return decodeURIComponent(value)
  } catch {
    return unescape(value)
  }
}

/**
 * @param {string} value
 * @returns {string}
 */
function encodeMailtoComponent (value) {
  return encodeURIComponent(decodeMailtoComponent(value))
    .replace(/%40/gi, '@')
    .replace(/%27/g, "'")
}

/**
 * @param {string} domain
 * @returns {string}
 */
function domainToASCII (domain) {
  try {
    return new URL('http://' + domain).hostname
  } catch {
    return domain
  }
}

/**
 * @param {string} value
 * @returns {string}
 */
function encodeMailtoAddress (value) {
  const address = decodeMailtoComponent(value)
  const at = address.lastIndexOf('@')
  if (at === -1) {
    return encodeURIComponent(address).replace(/%27/g, "'")
  }

  return encodeURIComponent(address.slice(0, at)).replace(/%27/g, "'") +
    '@' +
    domainToASCII(address.slice(at + 1).toLowerCase())
}

/**
 * @param {string} value
 * @param {import('../types/index').Options} options
 * @returns {string}
 */
function normalizeMailtoAddress (value, options) {
  const address = decodeMailtoComponent(value)
  const at = address.lastIndexOf('@')
  if (at === -1) {
    return address
  }

  const local = address.slice(0, at)
  const domain = address.slice(at + 1).toLowerCase()
  return local + '@' + (options.unicodeSupport ? domain : domainToASCII(domain))
}

/** @type {SchemeFn} */
function mailtoParse (mailtoComponent, options) {
  const to = []
  if (mailtoComponent.path) {
    to.push(...mailtoComponent.path.split(',').filter(Boolean).map((address) => normalizeMailtoAddress(address, options)))
  }

  if (mailtoComponent.query) {
    const headers = {}
    for (const header of mailtoComponent.query.split('&')) {
      const separator = header.indexOf('=')
      const rawName = separator === -1 ? header : header.slice(0, separator)
      const rawValue = separator === -1 ? '' : header.slice(separator + 1)
      const name = decodeMailtoComponent(rawName)
      const value = decodeMailtoComponent(rawValue)
      const lowerName = name.toLowerCase()

      if (lowerName === 'to') {
        to.push(...value.split(',').filter(Boolean).map((address) => normalizeMailtoAddress(address, options)))
      } else if (lowerName === 'subject') {
        mailtoComponent.subject = value
      } else if (lowerName === 'body') {
        mailtoComponent.body = value
      } else if (name) {
        headers[name] = value
      }
    }

    if (Object.keys(headers).length > 0) {
      mailtoComponent.headers = headers
    }
  }

  if (to.length > 0) {
    mailtoComponent.to = to
  }

  mailtoComponent.path = undefined
  mailtoComponent.query = undefined
  return mailtoComponent
}

/** @type {SchemeFn} */
function mailtoSerialize (mailtoComponent) {
  const headers = Object.assign({}, mailtoComponent.headers)
  if (mailtoComponent.subject !== undefined) {
    headers.subject = mailtoComponent.subject
  }
  if (mailtoComponent.body !== undefined) {
    headers.body = mailtoComponent.body
  }

  if (Array.isArray(mailtoComponent.to) && mailtoComponent.to.length > 0) {
    mailtoComponent.path = mailtoComponent.to.map(encodeMailtoAddress).join(',')
  }

  const query = Object.keys(headers).map((name) => {
    return encodeMailtoComponent(name) + '=' + encodeMailtoComponent(String(headers[name]))
  })
  if (query.length > 0) {
    mailtoComponent.query = query.join('&')
  }

  mailtoComponent.fragment = undefined
  return mailtoComponent
}

const http = /** @type {SchemeHandler} */ ({
  scheme: 'http',
  domainHost: true,
  parse: httpParse,
  serialize: httpSerialize
})

const https = /** @type {SchemeHandler} */ ({
  scheme: 'https',
  domainHost: http.domainHost,
  parse: httpParse,
  serialize: httpSerialize
})

const ws = /** @type {SchemeHandler} */ ({
  scheme: 'ws',
  domainHost: true,
  parse: wsParse,
  serialize: wsSerialize
})

const wss = /** @type {SchemeHandler} */ ({
  scheme: 'wss',
  domainHost: ws.domainHost,
  parse: ws.parse,
  serialize: ws.serialize
})

const urn = /** @type {SchemeHandler} */ ({
  scheme: 'urn',
  parse: urnParse,
  serialize: urnSerialize,
  skipNormalize: true
})

const urnuuid = /** @type {SchemeHandler} */ ({
  scheme: 'urn:uuid',
  parse: urnuuidParse,
  serialize: urnuuidSerialize,
  skipNormalize: true
})

const mailto = /** @type {SchemeHandler} */ ({
  scheme: 'mailto',
  parse: mailtoParse,
  serialize: mailtoSerialize,
  skipNormalize: true,
  skipEscape: true
})

const SCHEMES = /** @type {Record<SchemeName, SchemeHandler>} */ ({
  http,
  https,
  ws,
  wss,
  urn,
  'urn:uuid': urnuuid,
  mailto
})

Object.setPrototypeOf(SCHEMES, null)

/**
 * @param {string|undefined} scheme
 * @returns {SchemeHandler|undefined}
 */
function getSchemeHandler (scheme) {
  return (
    scheme && (
      SCHEMES[/** @type {SchemeName} */ (scheme)] ||
      SCHEMES[/** @type {SchemeName} */(scheme.toLowerCase())])
  ) ||
    undefined
}

module.exports = {
  wsIsSecure,
  SCHEMES,
  isValidSchemeName,
  getSchemeHandler,
}
