// fast-uri/lib/schemes.js
'use strict'

const { domainToASCII } = require('node:url')
const { isUUID } = require('./utils')
const URN_REG = /([\da-z][\d\-a-z]{0,31}):((?:[\w!$'()*+,\-.:;=@]|%[\da-f]{2})+)/iu
const PCT_ENCODED = /%[\da-f]{2}/giu
const UNRESERVED = /^[\da-z\-._~]$/iu
const NOT_LOCAL_PART = /[^A-Za-z0-9!$%'()*+,\-.<>^_`a-z{|}~]/gu
const NOT_HFIELD = /[^A-Za-z0-9\-._~!$'()*+,;:@]/gu

const supportedSchemeNames = /** @type {const} */ (['http', 'https', 'ws',
  'wss', 'urn', 'urn:uuid', 'mailto'])

/** @typedef {supportedSchemeNames[number]} SchemeName */

/**
 * @param {string} name
 * @returns {name is SchemeName}
 */
function isValidSchemeName (name) {
  return supportedSchemeNames.indexOf(/** @type {*} */ (name)) !== -1
}

/**
 * @callback SchemeFn
 * @param {import('../types/index').URIComponent} component
 * @param {import('../types/index').Options} options
 * @returns {import('../types/index').URIComponent}
 */

/**
 * @typedef {Object} SchemeHandler
 * @property {SchemeName} scheme - The scheme name.
 * @property {boolean} [domainHost] - Indicates if the scheme supports domain hosts.
 * @property {SchemeFn} parse - Function to parse the URI component for this scheme.
 * @property {SchemeFn} serialize - Function to serialize the URI component for this scheme.
 * @property {boolean} [skipNormalize] - Indicates if normalization should be skipped for this scheme.
 * @property {boolean} [skipEscape] - Indicates if path escaping should be skipped for this scheme.
 * @property {boolean} [absolutePath] - Indicates if the scheme uses absolute paths.
 * @property {boolean} [unicodeSupport] - Indicates if the scheme supports Unicode.
 */

/**
 * @param {import('../types/index').URIComponent} wsComponent
 * @returns {boolean}
 */
function wsIsSecure (wsComponent) {
  if (wsComponent.secure === true) {
    return true
  } else if (wsComponent.secure === false) {
    return false
  } else if (wsComponent.scheme) {
    return (
      wsComponent.scheme.length === 3 &&
      (wsComponent.scheme[0] === 'w' || wsComponent.scheme[0] === 'W') &&
      (wsComponent.scheme[1] === 's' || wsComponent.scheme[1] === 'S') &&
      (wsComponent.scheme[2] === 's' || wsComponent.scheme[2] === 'S')
    )
  } else {
    return false
  }
}

/** @type {SchemeFn} */
function httpParse (component) {
  if (!component.host) {
    component.error = component.error || 'HTTP URIs must have a host.'
  }

  return component
}

/** @type {SchemeFn} */
function httpSerialize (component) {
  const secure = String(component.scheme).toLowerCase() === 'https'

  // normalize the default port
  if (component.port === (secure ? 443 : 80) || component.port === '') {
    component.port = undefined
  }

  // normalize the empty path
  if (!component.path) {
    component.path = '/'
  }

  // NOTE: We do not parse query strings for HTTP URIs
  // as WWW Form Url Encoded query strings are part of the HTML4+ spec,
  // and not the HTTP spec.

  return component
}

/** @type {SchemeFn} */
function wsParse (wsComponent) {
// indicate if the secure flag is set
  wsComponent.secure = wsIsSecure(wsComponent)

  // construct resouce name
  wsComponent.resourceName = (wsComponent.path || '/') + (wsComponent.query ? '?' + wsComponent.query : '')
  wsComponent.path = undefined
  wsComponent.query = undefined

  return wsComponent
}

/** @type {SchemeFn} */
function wsSerialize (wsComponent) {
// normalize the default port
  if (wsComponent.port === (wsIsSecure(wsComponent) ? 443 : 80) || wsComponent.port === '') {
    wsComponent.port = undefined
  }

  // ensure scheme matches secure flag
  if (typeof wsComponent.secure === 'boolean') {
    wsComponent.scheme = (wsComponent.secure ? 'wss' : 'ws')
    wsComponent.secure = undefined
  }

  // reconstruct path from resource name
  if (wsComponent.resourceName) {
    const [path, query] = wsComponent.resourceName.split('?')
    wsComponent.path = (path && path !== '/' ? path : undefined)
    wsComponent.query = query
    wsComponent.resourceName = undefined
  }

  // forbid fragment component
  wsComponent.fragment = undefined

  return wsComponent
}

/** @type {SchemeFn} */
function urnParse (urnComponent, options) {
  if (!urnComponent.path) {
    urnComponent.error = 'URN can not be parsed'
    return urnComponent
  }
  const matches = urnComponent.path.match(URN_REG)
  if (matches) {
    const scheme = options.scheme || urnComponent.scheme || 'urn'
    urnComponent.nid = matches[1].toLowerCase()
    urnComponent.nss = matches[2]
    const urnScheme = `${scheme}:${options.nid || urnComponent.nid}`
    const schemeHandler = getSchemeHandler(urnScheme)
    urnComponent.path = undefined

    if (schemeHandler) {
      urnComponent = schemeHandler.parse(urnComponent, options)
    }
  } else {
    urnComponent.error = urnComponent.error || 'URN can not be parsed.'
  }

  return urnComponent
}

/** @type {SchemeFn} */
function urnSerialize (urnComponent, options) {
  if (urnComponent.nid === undefined) {
    throw new Error('URN without nid cannot be serialized')
  }
  const scheme = options.scheme || urnComponent.scheme || 'urn'
  const nid = urnComponent.nid.toLowerCase()
  const urnScheme = `${scheme}:${options.nid || nid}`
  const schemeHandler = getSchemeHandler(urnScheme)

  if (schemeHandler) {
    urnComponent = schemeHandler.serialize(urnComponent, options)
  }

  const uriComponent = urnComponent
  const nss = urnComponent.nss
  uriComponent.path = `${nid || options.nid}:${nss}`

  options.skipEscape = true
  return uriComponent
}

/** @type {SchemeFn} */
function urnuuidParse (urnComponent, options) {
  const uuidComponent = urnComponent
  uuidComponent.uuid = uuidComponent.nss
  uuidComponent.nss = undefined

  if (!options.tolerant && (!uuidComponent.uuid || !isUUID(uuidComponent.uuid))) {
    uuidComponent.error = uuidComponent.error || 'UUID is not valid.'
  }

  return uuidComponent
}

/** @type {SchemeFn} */
function urnuuidSerialize (uuidComponent) {
  const urnComponent = uuidComponent
  // normalize UUID
  urnComponent.nss = (uuidComponent.uuid || '').toLowerCase()
  return urnComponent
}

/**
 * @param {string} value
 * @returns {string}
 */
function decodeURIComponentSafe (value) {
  try {
    return decodeURIComponent(value)
  } catch {
    return value
  }
}

/**
 * @param {string} value
 * @returns {string}
 */
function normalizePctEncoding (value) {
  return value.replace(PCT_ENCODED, (encoded) => {
    const decoded = decodeURIComponentSafe(encoded)
    return UNRESERVED.test(decoded) ? decoded : encoded.toUpperCase()
  })
}

/**
 * @param {string} value
 * @param {RegExp} unsafe
 * @returns {string}
 */
function encodeMailtoComponent (value, unsafe) {
  return normalizePctEncoding(value).replace(unsafe, encodeURIComponent)
}

/**
 * @param {unknown} value
 * @returns {Array<unknown>}
 */
function toArray (value) {
  if (value === undefined || value === null) {
    return []
  }
  return Array.isArray(value) ? value : [value]
}

/**
 * @param {string} address
 * @param {import('../types/index').URIComponent} component
 * @param {import('../types/index').Options} options
 * @returns {string}
 */
function normalizeMailtoAddress (address, component, options) {
  const at = address.lastIndexOf('@')
  if (at === -1) {
    return decodeURIComponentSafe(address)
  }

  const local = decodeURIComponentSafe(address.slice(0, at))
  const domain = decodeURIComponentSafe(address.slice(at + 1)).toLowerCase()

  if (options.unicodeSupport) {
    return `${local}@${domain}`
  }

  try {
    return `${local}@${domainToASCII(domain)}`
  } catch (e) {
    component.error = component.error || "Email address's domain name can not be converted to ASCII: " + e
    return `${local}@${domain}`
  }
}

/** @type {SchemeFn} */
function mailtoParse (mailtoComponent, options) {
  const to = mailtoComponent.to = mailtoComponent.path ? mailtoComponent.path.split(',') : []
  mailtoComponent.path = undefined

  if (mailtoComponent.query) {
    const headers = {}
    let hasUnknownHeaders = false
    const hfields = mailtoComponent.query.split('&')

    for (let i = 0; i < hfields.length; i++) {
      const [name, value = ''] = hfields[i].split('=')
      switch (name) {
        case 'to':
          to.push(...value.split(','))
          break
        case 'subject':
          mailtoComponent.subject = decodeURIComponentSafe(value)
          break
        case 'body':
          mailtoComponent.body = decodeURIComponentSafe(value)
          break
        default:
          hasUnknownHeaders = true
          headers[decodeURIComponentSafe(name)] = decodeURIComponentSafe(value)
          break
      }
    }

    if (hasUnknownHeaders) {
      mailtoComponent.headers = headers
    }
  }

  mailtoComponent.query = undefined
  for (let i = 0; i < to.length; i++) {
    to[i] = normalizeMailtoAddress(to[i], mailtoComponent, options)
  }

  return mailtoComponent
}

/** @type {SchemeFn} */
function mailtoSerialize (mailtoComponent, options) {
  const to = toArray(mailtoComponent.to)

  if (to.length) {
    for (let i = 0; i < to.length; i++) {
      const toAddr = String(to[i])
      const at = toAddr.lastIndexOf('@')
      if (at === -1) {
        to[i] = encodeMailtoComponent(toAddr, NOT_LOCAL_PART)
        continue
      }

      const local = encodeMailtoComponent(toAddr.slice(0, at), NOT_LOCAL_PART)
      let domain = toAddr.slice(at + 1)

      if (!options.unicodeSupport) {
        try {
          domain = domainToASCII(decodeURIComponentSafe(domain).toLowerCase())
        } catch (e) {
          mailtoComponent.error = mailtoComponent.error || "Email address's domain name can not be converted to ASCII: " + e
        }
      } else {
        domain = decodeURIComponentSafe(domain).toLowerCase()
      }

      to[i] = `${local}@${domain}`
    }

    mailtoComponent.path = to.join(',')
  }

  const headers = mailtoComponent.headers || {}
  if (mailtoComponent.subject !== undefined) {
    headers.subject = mailtoComponent.subject
  }
  if (mailtoComponent.body !== undefined) {
    headers.body = mailtoComponent.body
  }

  const fields = []
  for (const name in headers) {
    fields.push(
      encodeMailtoComponent(name, NOT_HFIELD) +
      '=' +
      encodeMailtoComponent(String(headers[name]), NOT_HFIELD)
    )
  }

  if (fields.length) {
    mailtoComponent.query = fields.join('&')
  }

  options.skipEscape = true
  return mailtoComponent
}

const http = /** @type {SchemeHandler} */ ({
  scheme: 'http',
  domainHost: true,
  parse: httpParse,
  serialize: httpSerialize
})

const https = /** @type {SchemeHandler} */ ({
  scheme: 'https',
  domainHost: http.domainHost,
  parse: httpParse,
  serialize: httpSerialize
})

const ws = /** @type {SchemeHandler} */ ({
  scheme: 'ws',
  domainHost: true,
  parse: wsParse,
  serialize: wsSerialize
})

const wss = /** @type {SchemeHandler} */ ({
  scheme: 'wss',
  domainHost: ws.domainHost,
  parse: ws.parse,
  serialize: ws.serialize
})

const urn = /** @type {SchemeHandler} */ ({
  scheme: 'urn',
  parse: urnParse,
  serialize: urnSerialize,
  skipNormalize: true
})

const urnuuid = /** @type {SchemeHandler} */ ({
  scheme: 'urn:uuid',
  parse: urnuuidParse,
  serialize: urnuuidSerialize,
  skipNormalize: true
})

const mailto = /** @type {SchemeHandler} */ ({
  scheme: 'mailto',
  parse: mailtoParse,
  serialize: mailtoSerialize,
  skipNormalize: true,
  skipEscape: true
})

const SCHEMES = /** @type {Record<SchemeName, SchemeHandler>} */ ({
  http,
  https,
  ws,
  wss,
  urn,
  'urn:uuid': urnuuid,
  mailto
})

Object.setPrototypeOf(SCHEMES, null)

/**
 * @param {string|undefined} scheme
 * @returns {SchemeHandler|undefined}
 */
function getSchemeHandler (scheme) {
  return (
    scheme && (
      SCHEMES[/** @type {SchemeName} */ (scheme)] ||
      SCHEMES[/** @type {SchemeName} */(scheme.toLowerCase())])
  ) ||
    undefined
}

module.exports = {
  wsIsSecure,
  SCHEMES,
  isValidSchemeName,
  getSchemeHandler,
}
